### PR TITLE
[FLINK-21151] Extract common full-snapshot writer from RocksDB full-snapshot strategy

### PR DESCRIPTION
**Pull request by aljoscha**
As described in FLIP-41, the RocksDB full-snapshot format will serve as the common, unified savepoint format. We need to extract the common parts and make them reusable by other state backends.

The commits themselves have good description of what's going on, it's easiest to follow the commits one by one instead of reviewing the final results.

## Verifying this change

This is a pure refactoring and should be covered by existing tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/FullSnapshotUtil.java [hunk 1]
```diff
@@ -16,13 +16,13 @@
...
- * Utility methods and constants around RocksDB creating and restoring snapshots for {@link
- * org.apache.flink.contrib.streaming.state.RocksDBKeyedStateBackend}.
+ * Utility methods and constants around creating and restoring full snapshots using {@link
+ * FullSnapshotAsyncWriter}.
  */
-public class RocksSnapshotUtil {
+public class FullSnapshotUtil {
 
     /** File suffix of sstable files. */
     public static final String SST_FILE_SUFFIX = ".sst";
```
Comment (dawidwys) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/FullSnapshotUtil.java:28]: nit: This constant is Rocks specific, isn't it?

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/KeyValueStateIterator.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+     * Check if the iterator is still valid. Getters like {@link #key()}, {@link #value()}, etc. as
+     * well as {@link #next()} should only be called if valid returned true. Should be checked after
+     * each call to {@link #next()} before accessing iterator state.
+     *
+     * @return True iff this iterator is valid.
+     */
+    boolean isValid();
+
+    @Override
+    void close();
```
Comment (dawidwys) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/KeyValueStateIterator.java:72]: Do we want to override it without an exception?

File: flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/iterator/RocksSingleStateIterator.java [hunk 3]
```diff
@@ -23,11 +23,13 @@
...
 import javax.annotation.Nonnull;
 
+import java.io.Closeable;
+
 /**
  * Wraps a RocksDB iterator to cache it's current key and assigns an id for the key/value state to
  * the iterator. Used by {@link RocksStatesPerKeyGroupMergeIterator}.
  */
-class RocksSingleStateIterator implements AutoCloseable {
+class RocksSingleStateIterator implements Closeable {
```
Comment (dawidwys) [inline flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/iterator/RocksSingleStateIterator.java:32]: is this change necessary? The `AutoCloseable` is the newer interface.

File: flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/iterator/RocksStatesPerKeyGroupMergeIterator.java [hunk 4]
```diff
@@ -106,7 +118,8 @@ public void next() {
                 detectNewKeyGroup(oldKey);
             }
         } else {
-            IOUtils.closeQuietly(rocksIterator);
+            IOUtils.closeQuietly(currentSubIterator);
+            closeableRegistry.unregisterCloseable(currentSubIterator);
```
Comment (dawidwys) [inline flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/iterator/RocksStatesPerKeyGroupMergeIterator.java:122]: Usually the pattern is:

```
            if (closeableRegistry.unregisterCloseable(currentSubIterator)) {
                IOUtils.closeQuietly(currentSubIterator);
            }
```
BTW, do we even need to unregister it from the registry? I think it won't harm, worst case it will be closed twice (which was the case previously as well).

File: flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/iterator/RocksStatesPerKeyGroupMergeIterator.java [hunk 5]
```diff
@@ -132,10 +145,14 @@ public void next() {
             final RocksIteratorWrapper rocksIterator = rocksIteratorWithKVStateId.f0;
             rocksIterator.seekToFirst();
             if (rocksIterator.isValid()) {
-                iteratorPriorityQueue.offer(
-                        new RocksSingleStateIterator(rocksIterator, rocksIteratorWithKVStateId.f1));
+                RocksSingleStateIterator wrappingIterator =
+                        new RocksSingleStateIterator(rocksIterator, rocksIteratorWithKVStateId.f1);
+                iteratorPriorityQueue.offer(wrappingIterator);
+                closeableRegistry.registerCloseable(wrappingIterator);
+                closeableRegistry.unregisterCloseable(rocksIterator);
```
Comment (dawidwys) [inline flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/iterator/RocksStatesPerKeyGroupMergeIterator.java:152]: Same as above. Do we care about unregistering here?

File: flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/snapshot/RocksDBFullSnapshotResources.java [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+
+        try {
+            ReadOptions readOptions = new ReadOptions();
+            closeableRegistry.registerCloseable(readOptions::close);
+            readOptions.setSnapshot(snapshot);
+
+            List<Tuple2<RocksIteratorWrapper, Integer>> kvStateIterators =
+                    createKVStateIterators(readOptions);
+
+            for (Tuple2<RocksIteratorWrapper, Integer> iter : kvStateIterators) {
```
Comment (dawidwys) [inline flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/snapshot/RocksDBFullSnapshotResources.java:118]: nit: We could do that in a single pass when creating iterators.

**Review by dawidwys — COMMENTED** [review]
Nice job! We're getting closer.

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/FullSnapshotUtil.java (hunk 1 above)
Comment (aljoscha) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/FullSnapshotUtil.java:28]: True, I will have to move some of them back to a Rocks-specific `RocksSnapshotUtil`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/KeyValueStateIterator.java (hunk 2 above)
Comment (aljoscha) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/KeyValueStateIterator.java:72]: I kept this from `RocksStatesPerKeyGroupMergeIterator`, which also overrode `close()` without exception. And some of the code that uses it expects it not to throw.

File: flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/iterator/RocksSingleStateIterator.java (hunk 3 above)
Comment (aljoscha) [reply flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/iterator/RocksSingleStateIterator.java:32]: I made it `Closeable` because the `RocksStatesPerKeyGroupMergeIterator` now uses a `CloseableRegistry` which requires things to be `Closeable`. 

Alternatively, I could change `CloseableRegistry` to take `AutoCloseables` but I felt that would be the bigger change.

File: flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/iterator/RocksStatesPerKeyGroupMergeIterator.java (hunk 4 above)
Comment (aljoscha) [reply flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/iterator/RocksStatesPerKeyGroupMergeIterator.java:122]: Thanks, I will use this pattern! And yes, it's actually important to unregister, we have tests that verify that exactly the right amount of `close()` calls happen on the Rocks classes. I was surprised myself. 🎊

File: flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/iterator/RocksStatesPerKeyGroupMergeIterator.java (hunk 5 above)
Comment (aljoscha) [reply flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/iterator/RocksStatesPerKeyGroupMergeIterator.java:152]: See answer above.

File: flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/snapshot/RocksDBFullSnapshotResources.java (hunk 6 above)
Comment (aljoscha) [reply flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/snapshot/RocksDBFullSnapshotResources.java:118]: fixing

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/FullSnapshotAsyncWriter.java [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+import static org.apache.flink.runtime.state.FullSnapshotUtil.hasMetaDataFollowsFlag;
+import static org.apache.flink.runtime.state.FullSnapshotUtil.setMetaDataFollowsFlagInKey;
+
+/**
+ * An asynchronous writer that can write a full snapshot/savepoint from a {@link
+ * FullSnapshotResources}.
+ *
+ * @param <K> type of the backend keys.
+ */
+public class FullSnapshotAsyncWriter<K>
```
[the remark this answers is not in the document]
Comment (aljoscha) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/FullSnapshotAsyncWriter.java:44]: Can do, I have two thoughts on it:
 - the `RocksFullSnapshotStrategy` still uses this, and not just for savepoints
 - my latest commit introduces a `SavepointSnapshotStrategy` that uses this. In my head the `FullSnapshotAsyncWriter` is an implementation detail that happens to also be used for savepoints

Agreed that `Full` is not very clear here. Also, I'm not at all set on the naming.

File: flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/iterator/RocksSingleStateIterator.java (hunk 3 above)
Comment (dawidwys) [reply flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/iterator/RocksSingleStateIterator.java:32]: got it, I think that's fine

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/FullSnapshotAsyncWriter.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (dawidwys) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/FullSnapshotAsyncWriter.java:44]: Got it. I think it's fine to leave it as is then.

File: flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/snapshot/RocksSnapshotUtil.java [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+package org.apache.flink.contrib.streaming.state.snapshot;
+
+/**
+ * Utility methods and constants around RocksDB creating and restoring snapshots for {@link
+ * org.apache.flink.contrib.streaming.state.RocksDBKeyedStateBackend}.
+ */
+public class RocksSnapshotUtil {
+
+    /** File suffix of sstable files. */
+    public static final String SST_FILE_SUFFIX = ".sst";
```
Comment (dawidwys) [inline flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/snapshot/RocksSnapshotUtil.java]: Shall we leave just the `SST_FILE_SUFFIX` constant here?

**Review by dawidwys — APPROVED** [review]
Thanks for the update. I think the DRAFT commit makes sense! I had a last small comment and after that I think it's good to be merged.

File: flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/snapshot/RocksSnapshotUtil.java (hunk 8 above)
Comment (aljoscha) [reply flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/snapshot/RocksSnapshotUtil.java]: Ah damn, I meant to remove the other ones, yes.

**Comment by aljoscha** [comment]
merged! Thanks for the reviews, Dawid!
